### PR TITLE
implement json_object

### PR DESCRIPTION
**Pull request by valkrypton**
added support for `json_object` under [#4216](https://github.com/diesel-rs/diesel/issues/4216)

### CODE REVIEW

**Comment by valkrypton** [comment]
@weiznich  what value should i pass it in `auto_type.rs` as `pg_extras::array` is of type `Array<Integer>`?

**Comment by weiznich** [comment]
@valkrypton I would suggest to just define a new column with the type `Array<Text>` in the pg_extras table.

File: diesel/src/pg/expression/functions.rs [hunk 1]
```diff
@@ -1423,3 +1423,38 @@ define_sql_function! {
...
+    /// #     use diesel::dsl::json_object;
+    /// #     use diesel::sql_types::{Json};
+    /// #     use serde_json::Value;
+    /// #     let connection = &mut establish_connection();
+    /// let json = diesel::select(json_object(vec!["hello","world"])).get_result::<Value>(connection)?;
+    /// let expected:Value = serde_json::from_str(r#"{"hello":"world"}"#).unwrap();
+    /// assert_eq!(expected,json);
+    ///
+    /// let json = diesel::select(json_object(vec!["hello","world","John","Doe"])).get_result::<Value>(connection)?;
+    /// let expected:Value = serde_json::from_str(r#"{"hello":"world","John":"Doe"}"#).unwrap();
```
Comment (guissalustiano) [inline diesel/src/pg/expression/functions.rs:1450]: Could we use the serde_json::json! macro instead to construct the expected values?

File: diesel/src/pg/expression/functions.rs [hunk 2]
```diff
@@ -1423,3 +1423,38 @@ define_sql_function! {
...
+    /// assert_eq!(expected,json);
+    ///
+    /// let json = diesel::select(json_object(vec!["hello","world","John"])).get_result::<Value>(connection);
+    /// assert!(json.is_err());
+    ///
+    /// #     Ok(())
+    /// # }
+    /// ```
+    fn json_object(text_array:Array<Text>)->Json;
```
Comment (guissalustiano) [inline diesel/src/pg/expression/functions.rs:1459]: Array are always generated with nullable elements in diesel which can make hard to use this function. I think this functions should be able to accept `Nullable<Array<Nullable<Text>>`, maybe using a helper trait to make simple to call it directly.

File: diesel/src/pg/expression/functions.rs [hunk 3]
```diff
@@ -1423,3 +1423,38 @@ define_sql_function! {
     /// ```
     fn array_upper<Arr: ArrayOrNullableArray + SingleValue>(array: Arr, dimension: Integer) -> Nullable<Integer>;
 }
+
+define_sql_function! {
```
Comment (guissalustiano) [inline diesel/src/pg/expression/functions.rs:1427]: This function also support a form `jsonb_object(keys text[], values text[])`, whitch takes keys and values pairwise from two separate arrays. Could you implement that too?

**Review by guissalustiano — CHANGES_REQUESTED** [review]
Thanks for open the PR!

I have some comments about the type signature and also request the other variant of this function

File: diesel/src/pg/expression/functions.rs (hunk 3 above)
Comment (valkrypton) [reply diesel/src/pg/expression/functions.rs:1427]: yes sure ill implement that too

**Comment by valkrypton** [comment]
@guissalustiano can you take a look at the type signature now?

File: diesel/src/pg/expression/functions.rs (hunk 2 above)
Comment (weiznich) [reply diesel/src/pg/expression/functions.rs]: I would just accept both variants here. That might require an additional helper trait that is implemented for both `Array<Text>` and `Array<Nullable<Text>>`.  Additionally what's about `Nullable<Array<_>>` here?

**Comment by guissalustiano** [comment]
Sure!
The signature `fn json_object(text_array:Nullable<Array<Text>>)->Json;`  is broken because if you call `SELECT json_object(NULL)` it returns `NULL`, but the function signatures don't tell you that.

A signature like `fn json_object(text_array:Nullable<Array<Text>>)->Nullable<Json>;` would fix that, but is annoying to have this extra `Nullable<_>` when you use that with non-nullable inputs. On [`to_json`](https://github.com/diesel-rs/diesel/pull/4229/files#diff-bf88b4900340a435caafd7c947e7b9a95e61ff158ef37596ab3262b3c754c45fR3702-R3714) we use a helper trait to handle this. 

We also would like to support an `Nullable<Array<Nullable<Text>>>` because diesel always generates columns as `Array<Nullable<_>>`, and so not supporting that would mean that `json_object` couldn't be used with a table, only literal values.

**Comment by valkrypton** [comment]
Thanks for the detailed explanation, i appreciate it.
The pr got messed up when i rebased this branch onto the updated master, i had no idea the previous commits would show up here as well. I tried to fix but couldn't figure it out, so if you could help me I'd appreciate it.

**Review by weiznich — APPROVED** [review]
Thanks for the update. The signature looks good now.

I just rebased this PR again to resolve the merge conflicts, but the history already looked right before that. So either it was a github issue before or you already fixed it :pray: 

I also added a `#[diagnostic::on_unimplemented]` attribute to the new helper trait to improve the compiler error message if some user passes in a wrong typed value.

**Pull request merged**